### PR TITLE
Refactoring suggestins

### PR DESCRIPTION
**Pull request by frip**
My first shot at a github pull-request.
More may follow...

### CODE REVIEW

**Comment by Orasund** [comment]
Hi, by removing test.sh you have broken the CI pipeline.

The reason for having a generic test.sh is so that the CI pipeline can be language independed. I intend to do more languages besides python

**Comment by Orasund** [comment]
Hi, please rebase. I've moved the CI code into a ci.sh file.

As far as i can tell adding pyproject.toml broke the CI because now pytest is not included in the requirements.txt

**Comment by frip** [comment]
Sorry, didn't see your answer in time! 
But I found something interesting (I think) in `diff_files.py` L77ff: did you know, that your `get_node()` is called with y=-1? (What I am doing here is not stricktly refactoring, because I'm not returning the first element in the line...)

File: diffFiles/src/diff_files.py [hunk 1]
```diff
@@ -3,87 +3,107 @@
...
+    common_indices: list[Position] = []
+    if not nodes.nodes:
+        return common_indices
 
-    pos: tuple[int, int] = (len(nodes) - 1, len(nodes[0]) - 1)
+    pos = Position(len(nodes.nodes) - 1, len(nodes.nodes[0]) - 1)
 
-    def get_node(pos: tuple[int, int]) -> Node:
-        return nodes[pos[0]][pos[1]]
+    def get_node(p: Position) -> Node:
```
Comment (Orasund) [inline diffFiles/src/diff_files.py:77]: yeah, p can have negative values if the file is empty.
for the empty file, the algorithm breaks down. In particular, the matrix can't be walked, because the starting position does not exist.

I did not include that edge case, as it makes the program more complicated.

File: diffFiles/tests/test_diff_files.py [hunk 2]
```diff
@@ -23,3 +23,19 @@ def test_basic():
                 "(['\\n'], ['!'])"
                 "]\n")
     assert result.output == expected
+
+
+def test_diff():
+    os.chdir(os.path.dirname(__file__))
+    with open("data/file1") as f1, open("data/file2") as f2:
+        lines1 = f1.readlines()
+        lines2 = f2.readlines()
```
Comment (Orasund) [inline diffFiles/tests/test_diff_files.py:32]: If you want to include the edge cases, you can check at the beginning of the program if one of the lengths is zero and if so return all lines of the other file.

From that point onwards you can assume that the files are not empty.

File: diffFiles/src/diff_files.py [hunk 3]
```diff
@@ -1,106 +1,130 @@
...
+    nodes = Matrix.make(l1, l2)
     return nodes
 
 
-def compute_common_indices(nodes: list[list[Node]]) -> list[tuple[int, int]]:
-    pos: tuple[int, int] = (len(nodes) - 1, len(nodes[0]) - 1)
+def compute_common_indices(nodes: Matrix) -> list[Position]:
+    common_indices: list[Position] = []
+    if not nodes.nodes:
+        return common_indices
```
Comment (frip) [inline diffFiles/src/diff_files.py:73]: This looks after the "empty-file-case".

File: diffFiles/src/diff_files.py [hunk 4]
```diff
@@ -1,106 +1,130 @@
...
+
+    pos = Position(len(nodes.nodes) - 1, len(nodes.nodes[0]) - 1)
 
-    def get_node(pos: tuple[int, int]) -> Node:
-        return nodes[pos[0]][pos[1]]
+    def get_node(p: Position) -> Node:
+        result = nodes.get_node(p)
+        if result is None:
+            return Node(Position(0, 0), 0)
```
Comment (frip) [inline diffFiles/src/diff_files.py:80]: This here is my "tricky" fix for `p.y=-1`. It occurs in the "standard-test-case"!

Comment (Orasund) [reply diffFiles/src/diff_files.py:80]: I'm a bit puzzled. This is only possible if `len(nodes.nodes[0] - 1) = -1` or `next_pos.y = -1`.

The first case occurs if the second file is empty. Second case occurs if I forgot a guard inside `compute_node`. In particular, this would mean that the cast is not current.

File: diffFiles/src/diff_files.py [hunk 5]
```diff
@@ -1,106 +1,130 @@
...
+
+                # find the candidate with the bigger length
+                if node1 is None and node2 is None:
+                    next_pos = None
+                    length = 0
+                elif node2 is not None and (node1 is None or node1.length < node2.length):
+                    next_pos = candidate2
+                    length = node2.length
+                else:
+                    next_pos = candidate1
```
Comment (Orasund) [inline diffFiles/src/diff_files.py:50]: In order for `next_pos.y = -1` we would need to enter this branch with `node1 is None` (as there is a guard inside get_node). However, if `node1 is None`, then either branch 1 or 2 will be entered depending on `node1 is None`.

So im not seeing what should be wrong here.

**Comment by frip** [comment]
I saw it in the debugger! :-)

**Comment by Orasund** [comment]
Thanks a lot. I found the missing guard:
```
def compute_node(i1: int, i2: int) -> Node:
        next_pos: tuple[int, int] | None
        length: int

        if l1[i1] == l2[i2]:
            if i1 >= 0 and i2 >= 0:
                next_pos = (i1 - 1, i2 - 1)
                length = nodes[next_pos[0]][next_pos[1]].length + 1
            else:
                # this else branch was missing
                next_pos = None
                length = 0
```

**Comment by frip** [comment]
I tried adding your guard in the hope that I could then reduce the second closure `get_node()` to the (new) Matrix-method, but it didn't work... :-(

**Comment by Orasund** [comment]
Oh! bad luck :/

Hope it's no a dealbreaker. Sadly, I don't have the time to look more into it.
